### ISSUE TITLE
hwo to properly use kornia.enhance.histogram?

### ISSUE THREAD

**Issue by sbelharbi** [kornia/kornia#1217]
## 🐛 Bug

hi,
1. i am comparing `kornia.enhance.histogram` to `numpy.histogram` and `torch.histc`.
the last 2 give the same result, but kornia's histogram is different and unexpected...

2. also, in the [documentation](https://kornia.readthedocs.io/en/latest/enhance.html#kornia.enhance.histogram), it is not clear what is the impact of the `bandwidth` on the histogram? could you please make it clear. does it have something to do with the histogram precision. if yes, how (low/high values)?

## To Reproduce

all the 3 methods are required to produce histogram within the same range.

```python
import numpy as np
import kornia
from kornia.enhance import histogram

import torch


if __name__ == '__main__':
    print('torch:', torch.__version__)
    print('numpy:', np.__version__)
    print('kornia:', kornia.__version__)

    seed = 0
    torch.manual_seed(seed)
    np.random.seed(seed)
    torch.cuda.manual_seed(seed)
    torch.cuda.manual_seed_all(seed)

    device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
    nbins = 256
    a = torch.rand((1, 5))
    print(a)
    anp = a.numpy()
    nphist, bins = np.histogram(anp, bins=nbins, range=(anp.min(), anp.max()))
    print('np hist', nphist)

    a = a.to(device)
    thistc = torch.histc(a, bins=nbins, min=a.min(), max=a.max())
    print('torch hist', thistc)

    _bins = torch.linspace(a.min(), a.max(), 256).to(device)
    bandwith = torch.tensor(.9).to(device)

    khist = histogram(x=a, bins=_bins,
                      bandwidth=bandwith)
    print('kornia hist', khist[0])

    print('dif torch numpy: ', torch.abs(thistc - torch.from_numpy(nphist).to(
        device)).sum())



```

## Expected behavior

i expect the results to be close to torch and numpy and to be consistent.
the obtained histogram is not. all bins have values...
## Environment

```sh
Collecting environment information...
PyTorch version: 1.9.0
Is debug build: False
CUDA used to build PyTorch: 11.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.5 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.10

Python version: 3.7.10 (default, Feb 26 2021, 18:47:35)  [GCC 7.3.0] (64-bit runtime)
Python platform: Linux-4.15.0-122-generic-x86_64-with-debian-buster-sid
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: 
GPU 0: Tesla P100-PCIE-16GB
GPU 1: Tesla P100-PCIE-16GB

Nvidia driver version: 455.32.00
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.4.2
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] efficientnet-pytorch==0.7.0
[pip3] numpy==1.20.1
[pip3] torch==1.9.0
[pip3] torchvision==0.10.0
[conda] Could not collect
```

## Additional context

Obtained results:
```sh
torch: 1.9.0
numpy: 1.20.1
kornia: 0.5.5+9a70e35
tensor([[0.4963, 0.7682, 0.0885, 0.1320, 0.3074]])
np hist [1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1]
torch hist tensor([1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 1.], device='cuda:0')
kornia hist tensor([0.0038, 0.0039, 0.0039, 0.0039, 0.0039, 0.0039, 0.0039, 0.0039, 0.0039,
        0.0039, 0.0039, 0.0039, 0.0039, 0.0039, 0.0039, 0.0039, 0.0039, 0.0039,
        0.0039, 0.0039, 0.0039, 0.0039, 0.0039, 0.0039, 0.0039, 0.0039, 0.0039,
        0.0039, 0.0039, 0.0039, 0.0039, 0.0039, 0.0039, 0.0039, 0.0039, 0.0039,
        0.0039, 0.0039, 0.0039, 0.0039, 0.0039, 0.0040, 0.0040, 0.0040, 0.0040,
        0.0040, 0.0040, 0.0040, 0.0040, 0.0040, 0.0040, 0.0040, 0.0040, 0.0040,
        0.0040, 0.0040, 0.0040, 0.0040, 0.0040, 0.0040, 0.0040, 0.0040, 0.0040,
        0.0040, 0.0040, 0.0040, 0.0040, 0.0040, 0.0040, 0.0040, 0.0040, 0.0040,
        0.0040, 0.0040, 0.0040, 0.0040, 0.0040, 0.0040, 0.0040, 0.0040, 0.0040,
        0.0040, 0.0040, 0.0040, 0.0040, 0.0040, 0.0040, 0.0040, 0.0040, 0.0040,
        0.0040, 0.0040, 0.0040, 0.0040, 0.0040, 0.0040, 0.0040, 0.0040, 0.0040,
        0.0040, 0.0040, 0.0040, 0.0040, 0.0040, 0.0040, 0.0040, 0.0040, 0.0040,
        0.0040, 0.0040, 0.0040, 0.0040, 0.0040, 0.0040, 0.0040, 0.0040, 0.0040,
        0.0040, 0.0040, 0.0040, 0.0040, 0.0040, 0.0040, 0.0040, 0.0040, 0.0040,
        0.0040, 0.0040, 0.0040, 0.0040, 0.0040, 0.0040, 0.0040, 0.0040, 0.0040,
        0.0040, 0.0040, 0.0040, 0.0040, 0.0040, 0.0040, 0.0040, 0.0040, 0.0040,
        0.0040, 0.0040, 0.0040, 0.0040, 0.0040, 0.0040, 0.0040, 0.0040, 0.0040,
        0.0040, 0.0040, 0.0040, 0.0040, 0.0040, 0.0039, 0.0039, 0.0039, 0.0039,
        0.0039, 0.0039, 0.0039, 0.0039, 0.0039, 0.0039, 0.0039, 0.0039, 0.0039,
        0.0039, 0.0039, 0.0039, 0.0039, 0.0039, 0.0039, 0.0039, 0.0039, 0.0039,
        0.0039, 0.0039, 0.0039, 0.0039, 0.0039, 0.0039, 0.0039, 0.0039, 0.0039,
        0.0039, 0.0039, 0.0039, 0.0039, 0.0039, 0.0039, 0.0039, 0.0039, 0.0039,
        0.0039, 0.0038, 0.0038, 0.0038, 0.0038, 0.0038, 0.0038, 0.0038, 0.0038,
        0.0038, 0.0038, 0.0038, 0.0038, 0.0038, 0.0038, 0.0038, 0.0038, 0.0038,
        0.0038, 0.0038, 0.0038, 0.0038, 0.0038, 0.0038, 0.0038, 0.0038, 0.0038,
        0.0038, 0.0038, 0.0037, 0.0037, 0.0037, 0.0037, 0.0037, 0.0037, 0.0037,
        0.0037, 0.0037, 0.0037, 0.0037, 0.0037, 0.0037, 0.0037, 0.0037, 0.0037,
        0.0037, 0.0037, 0.0037, 0.0037, 0.0037, 0.0037, 0.0037, 0.0037, 0.0036,
        0.0036, 0.0036, 0.0036, 0.0036], device='cuda:0')
dif torch numpy:  tensor(0., device='cuda:0')
```

thanks

**Comment by edgarriba** [comment]
@sbelharbi thanks for reporting that - I think you are right, I'm playing with your example and something is wrong here -> https://colab.research.google.com/drive/1hGNcsWfCKA1T3ZnvVElDNcHZPnkRlQEP?usp=sharing

@bsuleymanov recently included a version for 2d images, maybe he can give some bits on it. Besides, I'm seeing that `kornia.enhance.histogram` doesn't have a proper regression test. We should add that at some point.

**Comment by bsuleymanov** [comment]
@edgarriba I had the same question during testing the implementation :) https://github.com/kornia/kornia/issues/790#issuecomment-886244845

@sbelharbi 
kornia.enhance.histogram method returns probabilities, not counts, thus we have 0.039, ..., values that sum to 1. image_histogram2d returns both histogram and probabilities. Also I found unwanted behavior of image_histogram2d in some cases: counts do not sum to n_points, this can be done with firstly calculating probabilities and then multiplying by n_points.

Bandwidth is a hyperparameter, but I have no answer how to find a proper value (except the default one) for it now. The default is (max - min) / n_bins. If you use Gaussian kernel, the large bandwidth will give you smooth unimodal approximation of your multimodal distribution, the small bandwidth results in multimodal approximation with sharp peaks.

I need some time to find a rule of thumb for choosing bandwidth value in general case of images.

**Comment by edgarriba** [comment]
@bsuleymanov thanks for the clarification. /cc @connorlee77 now I recall that you proposed the original implementation - could you give some tips here and how eventually we match (if possible) numpy/torch implementations ? /cc @ducha-aiki @lferraz any opinions here ?

**Comment by lyhyl** [comment]
@bsuleymanov @edgarriba I think it is not possible to make gaussian-function-based histogram can match numpy implement. Because sigma can only affect the width of gaussian distribution. Since the gaussian cannot "fill" a bin perfectly, it is not possible to match `np.hist`.

<img src="https://user-images.githubusercontent.com/3370369/133892881-dcb49899-ab81-47eb-84d9-31b57db5ae9c.png" alt="drawing" width="50%"/>

But the sigmoid function can.

<img src="https://discuss.pytorch.org/uploads/default/original/3X/4/e/4e5fd6cd9b94c10bde3bb87fff9fee1ca9a96fdf.png" alt="drawing" width="50%"/>

I adopted [this post](https://discuss.pytorch.org/t/differentiable-torch-histc/25865/4) with some modification (because kornia's histogram can have non-uniform bins).
Here is my implement: https://github.com/lyhyl/kornia/commit/12c1dc1c4e70fec103d88057451b09dc63573644
To test this implement:
```
import numpy as np
import torch

from kornia.enhance import histogram

seed = 0
torch.manual_seed(seed)
np.random.seed(seed)
torch.cuda.manual_seed(seed)
torch.cuda.manual_seed_all(seed)

device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
nbins = 256
a = torch.rand((1, 5))
print(a)

anp = a.numpy()
nphist, bins = np.histogram(anp, bins=nbins, range=(anp.min(), anp.max()))

a = a.to(device)
bins = torch.linspace(a.min(), a.max(), 256).to(device)

khist = histogram(a, bins, torch.tensor((a.max() - a.min()) / (nbins*3)).to(device), "gaussian")
diff = khist * np.prod(a.shape) - torch.from_numpy(nphist).to(device)
print('diff torch (gaussian) - numpy: ', torch.abs(diff).mean())

for i in range(4):
    khist = histogram(a, bins, torch.tensor(10**i).to(device), "sigmoid")
    diff = khist * np.prod(a.shape) - torch.from_numpy(nphist).to(device)
    print('diff torch (sigmoid) - numpy: ', torch.abs(diff).mean())
```
Output:
```
diff torch (gaussian) - numpy:  tensor(0.0031, device='cuda:0')
diff torch (sigmoid) - numpy:  tensor(0.0279, device='cuda:0')
diff torch (sigmoid) - numpy:  tensor(0.0017, device='cuda:0')
diff torch (sigmoid) - numpy:  tensor(8.4545e-10, device='cuda:0')
diff torch (sigmoid) - numpy:  tensor(0., device='cuda:0')
```
As shown above, this sigmoid-based implement has a good property:
When the bandwidth (sigma) approaches infinity, the difference approaches zero.
But the drawback is sigmoid may cause the vanishing gradient problem.

**Comment by bsuleymanov** [comment]
@lyhyl that's interesting, thanks! In the histogram loss paper author's use something like a triangular kernel to approximate histogram, I need to compare them to better understand.

To use sigmoid kernel in image_histogram2d, we just need to add
`
elif kernel == "sigmoid":
        kernel_values = 1 / (torch.exp(u) + torch.exp(-u))
`.

**Comment by edgarriba** [comment]
@lyhyl nice finding ! Worth to send a PR !

**Comment by lyhyl** [comment]
> @lyhyl nice finding ! Worth to send a PR !

Do we need to discuss it more deeply? In addition, I don't quite understand why we need `histogram` and `image_histogram2d` at the same time. Maybe we just need to flatten images before feed it into `histogram` and we will get a result consistent with `image_histogram2d`. Currently, `histogram` and `image_histogram2d` have different function signature, support different kernels, and `histogram` allows non-uniform bins while `image_histogram2d` only accept bins count as parameter. Please point out if I get something wrong.

**Comment by edgarriba** [comment]
I agree that both are very similar - I would unify. @connorlee77 was the first to introduce histrogram I believe. Any thoughts ?

**Comment by bsuleymanov** [comment]
We can use `image_histogram2d` instead of `torch.histc` for histogram equalization, for example. To use `torch.histc` we don't need to flatten input images.

About uniform bins: `image_histogram2d` accepts centers parameter, not only bins count.

**Comment by edgarriba** [comment]
@bsuleymanov let's do that - feel free to propose a PR

**Comment by connorlee77** [comment]
> @bsuleymanov thanks for the clarification. /cc @connorlee77 now I recall that you proposed the original implementation - could you give some tips here and how eventually we match (if possible) numpy/torch implementations ? /cc @ducha-aiki @lferraz any opinions here ?

Yeah, the kornia `histogram` function I implemented is a **differentiable** approximate of a histogram. It won't match the numpy or torch implementation exactly. I believe `bandwidth `is the smoothing parameter (higher value = less noise) of the gaussian kernel used in the kernel density estimation process.
